From 92ee52d4ad23e9f6b046febf6d87c36c3a266c1e Mon Sep 17 00:00:00 2001
From: dhineskumareswaran <dhineskumar.eswaran@celestica.com>
Date: Mon, 5 Dec 2022 12:17:20 +0000
Subject: [PATCH] Added shutdown path in spi_aspeed_smc driver

Summary:
Restoring the register back to default state by adding shutdown
path in spi-aspeed-smc driver.

Problem statement:
On the AST25XX chip, sometimes the flash gets stuck during the
reset state (reboot)

Root Cause:
Some timing issues are happening in flash reset logic.

Solution:
In reset path, restoring all FMC releated register back to default state.
---
 drivers/spi/spi-aspeed-smc.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/spi/spi-aspeed-smc.c b/drivers/spi/spi-aspeed-smc.c
index faf877e52bfe..a8e7b98ba7f4 100644
--- a/drivers/spi/spi-aspeed-smc.c
+++ b/drivers/spi/spi-aspeed-smc.c
@@ -46,6 +46,8 @@
 
 /* CEx Address Decoding Range Register */
 #define CE0_SEGMENT_ADDR_REG		0x30
+#define CE1_SEGMENT_ADDR_REG		0x34
+#define CE2_SEGMENT_ADDR_REG		0x38
 
 /* CEx Read timing compensation register */
 #define CE0_TIMING_COMPENSATION_REG	0x94
@@ -800,6 +802,20 @@ static int aspeed_spi_remove(struct platform_device *pdev)
 	return 0;
 }
 
+static void aspeed_spi_shutdown(struct platform_device *pdev)
+{
+	struct aspeed_spi *aspi = platform_get_drvdata(pdev);
+
+	/* restoring the register which we used before to default*/
+	writel(0x8000002A, aspi->regs + CONFIG_REG);
+	writel(0x700, aspi->regs + CE_CTRL_REG);
+	writel(0x0, aspi->regs + CE0_CTRL_REG);
+	writel(0x50400000, aspi->regs + CE0_SEGMENT_ADDR_REG);
+	writel(0x54500000, aspi->regs + CE1_SEGMENT_ADDR_REG);
+	writel(0x58540000, aspi->regs + CE2_SEGMENT_ADDR_REG);
+	writel(0x0, aspi->regs + CE0_TIMING_COMPENSATION_REG);
+}
+
 /*
  * AHB mappings
  */
@@ -1206,6 +1222,7 @@ MODULE_DEVICE_TABLE(of, aspeed_spi_matches);
 static struct platform_driver aspeed_spi_driver = {
 	.probe			= aspeed_spi_probe,
 	.remove			= aspeed_spi_remove,
+	.shutdown		= aspeed_spi_shutdown,
 	.driver	= {
 		.name		= DEVICE_NAME,
 		.of_match_table = aspeed_spi_matches,
-- 
2.25.1

